File: backend/app/Models/Product.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasOneThrough;
use Illuminate\Support\Str;

class Product extends Model
{
    protected $table = 'products';

    protected $fillable = [
        'product_type_id',
        'code',
        'name',
        'description',
        'price',
        'unit',
        'minimum_order',
        'sort_order',
        'featured',
        'is_active',
    ];

    protected $casts = [
        'price' => 'decimal:2',
        'minimum_order' => 'decimal:2',
        'specifications' => 'array',
        'sort_order' => 'integer',
        'featured' => 'boolean',
        'is_active' => 'boolean',
    ];

    public function productType(): BelongsTo
    {
        return $this->belongsTo(ProductType::class, 'product_type_id');
    }

    public function category(): HasOneThrough
    {
        return $this->hasOneThrough(
            ProductCategory::class,
            ProductType::class,
            'id',
            'id',
            'product_type_id',
            'category_id'
        );
    }

    /*
    |--------------------------------------------------------------------------
    | Scopes
    |--------------------------------------------------------------------------
    */

    public function scopeActive(Builder $query): Builder
    {
        return $query->where('is_active', true);
    }

    public function scopeInactive(Builder $query): Builder
    {
        return $query->where('is_active', false);
    }

    public function scopeFeatured(Builder $query): Builder
    {
        return $query->where('featured', true);
    }

    public function scopeNotFeatured(Builder $query): Builder
    {
        return $query->where('featured', false);
    }

    public function scopeByType(Builder $query, int $typeId): Builder
    {
        return $query->where('product_type_id', $typeId);
    }

    public function scopeByCategory(Builder $query, int $categoryId): Builder
    {
        return $query->whereHas('productType', function (Builder $q) use ($categoryId) {
            $q->where('category_id', $categoryId);
        });
    }

    public function scopeOrdered(Builder $query): Builder
    {
        return $query->orderBy('sort_order', 'asc')
            ->orderBy('name', 'asc');
    }

    public function scopeSearch(Builder $query, ?string $search): Builder
    {
        if (empty($search)) {
            return $query;
        }

        return $query->where(function (Builder $q) use ($search) {
            $q->where('name', 'like', "%{$search}%")
                ->orWhere('code', 'like', "%{$search}%")
                ->orWhere('description', 'like', "%{$search}%");
        });
    }

    public function scopePriceRange(Builder $query, ?float $min, ?float $max): Builder
    {
        if ($min !== null) {
            $query->where('price', '>=', $min);
        }

        if ($max !== null) {
            $query->where('price', '<=', $max);
        }

        return $query;
    }

    public function isActive(): bool
    {
        return $this->is_active === true;
    }

    public function isFeatured(): bool
    {
        return $this->featured === true;
    }

    public function toggleActive(): void
    {
        $this->update([
            'is_active' => !$this->is_active,
        ]);
    }

    public function toggleFeatured(): void
    {
        $this->update([
            'featured' => !$this->featured,
        ]);
    }

    public function activate(): void
    {
        $this->update(['is_active' => true]);
    }

    public function deactivate(): void
    {
        $this->update(['is_active' => false]);
    }

    public function markAsFeatured(): void
    {
        $this->update(['featured' => true]);
    }

    public function removeFeatured(): void
    {
        $this->update(['featured' => false]);
    }

    public function getFormattedPriceAttribute(): string
    {
        $price = (float) ($this->price ?? 0);
        return 'Rp ' . number_format($price, 0, ',', '.');
    }

    public function getPriceWithUnitAttribute(): string
    {
        return $this->formatted_price . ' / ' . ($this->unit ?? 'unit');
    }

    public function getCategoryNameAttribute(): string
    {
        return $this->productType?->category?->name ?? '-';
    }

    public function getTypeNameAttribute(): string
    {
        return $this->productType?->name ?? '-';
    }

    public static function generateCode(string $prefix = 'PRD'): string
    {
        do {
            $code = $prefix . '-' . strtoupper(Str::random(6));
        } while (self::where('code', $code)->exists());

        return $code;
    }

    public static function getByCategory(int $categoryId): Builder
    {
        return self::with(['productType.category'])
            ->byCategory($categoryId)
            ->active()
            ->ordered();
    }

    public static function getFeatured(int $limit = 10): Builder
    {
        return self::with(['productType.category'])
            ->featured()
            ->active()
            ->ordered()
            ->limit($limit);
    }

    protected static function boot()
    {
        parent::boot();

        static::creating(function (Product $product) {
            if (empty($product->code)) {
                $product->code = self::generateCode();
            }
        });
    }
}

File: app_mobile/lib/models/product.dart
class Product {
  final int id;
  final String? code;
  final String name;
  final String? description;
  final num price;
  final String unit;
  final num minimumOrder;
  final int sortOrder;
  final bool featured;
  final bool isActive;
  final int? productTypeId;
  final ProductType? productType;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Product({
    required this.id,
    this.code,
    required this.name,
    this.description,
    required this.price,
    this.unit = 'unit',
    this.minimumOrder = 1,
    this.sortOrder = 0,
    this.featured = false,
    this.isActive = true,
    this.productTypeId,
    this.productType,
    this.createdAt,
    this.updatedAt,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] ?? 0,
      code: json['code'],
      name: json['name'] ?? '',
      description: json['description'],
      price: (json['price'] is num)
          ? json['price']
          : num.tryParse(json['price']?.toString() ?? '0') ?? 0,
      unit: json['unit'] ?? 'unit',
      minimumOrder: (json['minimum_order'] is num)
          ? json['minimum_order']
          : num.tryParse(json['minimum_order']?.toString() ?? '1') ?? 1,
      sortOrder: json['sort_order'] is int
          ? json['sort_order']
          : int.tryParse(json['sort_order']?.toString() ?? '0') ?? 0,
      featured: json['featured'] == true ||
          json['featured'] == 1 ||
          json['featured'] == '1',
      isActive: json['is_active'] == true ||
          json['is_active'] == 1 ||
          json['is_active'] == '1',
      productTypeId: json['product_type_id'] is int
          ? json['product_type_id']
          : int.tryParse(json['product_type_id']?.toString() ?? ''),
      productType: json['product_type'] != null
          ? ProductType.fromJson(json['product_type'])
          : null,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'])
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.tryParse(json['updated_at'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'code': code,
      'name': name,
      'description': description,
      'price': price,
      'unit': unit,
      'minimum_order': minimumOrder,
      'sort_order': sortOrder,
      'featured': featured,
      'is_active': isActive,
      'product_type_id': productTypeId,
    };
  }
}

class ProductType {
  final int id;
  final String name;
  final int? categoryId;
  final Category? category;

  ProductType({
    required this.id,
    required this.name,
    this.categoryId,
    this.category,
  });

  factory ProductType.fromJson(Map<String, dynamic> json) {
    return ProductType(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      categoryId: json['category_id'] is int
          ? json['category_id']
          : int.tryParse(json['category_id']?.toString() ?? ''),
      category:
          json['category'] != null ? Category.fromJson(json['category']) : null,
    );
  }
}

class Category {
  final int id;
  final String name;

  Category({required this.id, required this.name});

  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
    );
  }
}
